pkg/controller/job/job_controller.go

```diff
@@ -496,12 +496,14 @@ func (jm *JobController) syncJob(key string) (bool, error) {
 	var failureMessage string
 
 	jobHaveNewFailure := failed > job.Status.Failed
+	exceedsBackoffLimit := jobHaveNewFailure && (int32(previousRetry)+1 > *job.Spec.BackoffLimit)
 
-	// check if the number of failed jobs increased since the last syncJob
-	if jobHaveNewFailure && (int32(previousRetry)+1 > *job.Spec.BackoffLimit) {
+	if exceedsBackoffLimit || pastBackoffLimitOnFailure(&job, pods) {
```

Comment:
Shouldn't it be 
```go
if exceedsBackoffLimit || (jobHaveNewFailure && pastBackoffLimitOnFailure(&job, pods)) 
```
?

Because you only check the pod restarts/job retries  when the job fails?

If so, please add a test for this case too.

Reply:
No, the current check is needed. `jobHaveNewFailure` verifies actual pod failures, which for `OnFailure` restart policy won't happen (unless the kubelet will kill the pod). So we always need to check these numbers as is.

Reply:
I think the test case I've added in this PR nicely covers it.